localstack/services/kinesis/kinesis_starter.py

```diff
@@ -1,214 +1,80 @@
 import logging
-import threading
-import time
+from typing import Optional
 
 from localstack import config
-from localstack.constants import MODULE_MAIN_PATH
-from localstack.services import install
-from localstack.services.infra import do_run, log_startup_message, start_proxy_for_service
-from localstack.services.kinesis import kinesis_listener
+from localstack.services.infra import log_startup_message, start_proxy_for_service
+from localstack.services.kinesis import kinesalite_server, kinesis_listener, kinesis_mock_server
 from localstack.utils.aws import aws_stack
-from localstack.utils.common import (
-    chmod_r,
-    get_free_tcp_port,
-    mkdir,
-    replace_in_file,
-    start_thread,
-    wait_for_port_open,
-)
+from localstack.utils.serving import Server
 
 LOG = logging.getLogger(__name__)
-
-# event to indicate that the kinesis backend service has stopped (the terminal command has returned)
-kinesis_stopped = threading.Event()
-
-# todo: will be replaced with plugin mechanism
-PROCESS_THREAD = None
-
-PORT_KINESIS_BACKEND = None
-
-
-def apply_patches_kinesalite():
-    files = [
-        "%s/kinesalite/validations/decreaseStreamRetentionPeriod.js",
-        "%s/kinesalite/validations/increaseStreamRetentionPeriod.js",
-    ]
-    for file_path in files:
-        file_path = file_path % install.INSTALL_DIR_NPM
-        replace_in_file("lessThanOrEqual: 168", "lessThanOrEqual: 8760", file_path)
-
-
-def start_kinesis(port=None, asynchronous=False, update_listener=None):
-    port = port or config.PORT_KINESIS
-    if config.KINESIS_PROVIDER == "kinesis-mock":
-        return start_kinesis_mock(
-            port=port, asynchronous=asynchronous, update_listener=update_listener
-        )
-    if config.KINESIS_PROVIDER == "kinesalite":
-        return start_kinesalite(
-            port=port, asynchronous=asynchronous, update_listener=update_listener
-        )
-    raise Exception('Unsupported Kinesis provider "%s"' % config.KINESIS_PROVIDER)
-
-
-def _run_proxy_and_command(cmd, port, backend_port, update_listener, asynchronous):
-    global PROCESS_THREAD
-    log_startup_message("Kinesis")
-    start_proxy_for_service("kinesis", port, backend_port, update_listener)
-
-    # TODO: generalize into service manager once it is introduced
-    LOG.debug("Running Kinesis startup command: %s", cmd)
-    try:
-        PROCESS_THREAD = do_run(cmd, asynchronous)
-    finally:
-        if asynchronous:
-
-            def _return_listener(*_):
-                try:
-                    ret_code = PROCESS_THREAD.result_future.result()
-                    if ret_code not in [0, None]:
-                        LOG.error("kinesis terminated with return code %s", ret_code)
-                finally:
-                    kinesis_stopped.set()
-
-            start_thread(_return_listener)
+_server: Optional[Server] = None  # server singleton
+
+
```

Comment:
I relied on `dynamdb_starter.py` heavily for refactoring this file. Parts of the code such as `_start_kinesis_helper()` and `restart_kinesis()` are nearly copy/pasted from there. To avoid further repetition, it might be worth considering either adding these as abstractions to the base `Server` class, or creating a new abstraction layer over `Server` to hold this logic (e.g. a `ServerStarter` class)